File: qetag.go
package main

import (
	"bytes"
	"crypto/sha1"
	"encoding/base64"
	"fmt"
	"io"
	"log"
	"os"
)

const (
	BLOCK_BITS = 22 // Indicate that the blocksize is 4M
)

// To get how many blocks does the file has.
func BlockCount(fsize int64) int {
	blockMask := int64((1 << BLOCK_BITS) - 1)
	return int((fsize + blockMask) >> BLOCK_BITS)
}

func CalSha1(r io.Reader) []byte {
	h := sha1.New()
	io.Copy(h, r)
	return h.Sum(nil)
}

func GetEtag(filename string) (etag string, err error) {
	f, err := os.Open(filename)
	defer f.Close()
	if err != nil {
		log.Println("Open file error : ", err)
		return
	}
	fi, err := f.Stat()
	if err != nil {
		log.Println(err)
		return
	}
	fsize := fi.Size()
	blockCnt := BlockCount(fsize)
	sha1Buf := make([]byte, 0, 21)

	if blockCnt <= 1 { // file size <= 4M
		sha1Buf = append(sha1Buf, 0x16)
		sha1Code := CalSha1(f)
		sha1Buf = append(sha1Buf, sha1Code...)
	} else { // file size > 4M
		sha1Buf = append(sha1Buf, 0x96)
		sha1BlockBuf := make([]byte, 0, blockCnt * 20)

		for i := 0; i < blockCnt; i ++ {
			body := io.LimitReader(f, 1 << BLOCK_BITS)
			sha1Code := CalSha1(body)
			sha1BlockBuf = append(sha1BlockBuf,sha1Code...)
		}
		tmpBuf := bytes.NewBuffer(sha1BlockBuf)
		sha1Final := CalSha1(tmpBuf)
		sha1Buf = append(sha1Buf, sha1Final...)
	}
	etag = base64.URLEncoding.EncodeToString(sha1Buf)
	return
}

func main() {

	usage := `usage : qetag <filename>`
	if len(os.Args) < 2 {
		fmt.Println(usage)
		return
	}
	etag, err := GetEtag(os.Args[1])
	if err != nil {
		log.Println("Calculate etag failed : ", err)
	}
	fmt.Println(etag)
}

File: qetag.php
<?php 

/**
 * 最新版本 在 https://github.com/qiniu/php-sdk/blob/master/src/Qiniu/Etag.php
 */
define('BLOCK_BITS', 22);
define('BLOCK_SIZE', 1 << BLOCK_BITS);


error_reporting(E_ERROR | E_WARNING | E_PARSE);


function PackArray($v, $a) {
    return call_user_func_array('pack', array_merge(array($v),(array)$a));
}


function BlockCount($fsize) {
    return (($fsize + (BLOCK_SIZE - 1)) >> BLOCK_BITS);
}


function URLSafeBase64Encode($str) // URLSafeBase64Encode
{
    $find = array('+', '/');
    $replace = array('-', '_');
    return str_replace($find, $replace, base64_encode($str));
}


function CalSha1($fhandler) {
    $fdata = fread($fhandler, BLOCK_SIZE);
    $sha1Str = sha1($fdata, true);
    $err = error_get_last();
    if ($err != null) {
        return array(null, $err);
    }
    $byteArray = unpack('C*', $sha1Str);
    return array($byteArray, null);
}


function GetEtag($filename) {
    if (!is_file($filename)) {
        $err = array ('message' => 'Can not open ' . $filename . ' as a file.');
        return array(null, $err);
    }
    $fhandler = fopen($filename, 'r');
    $err = error_get_last();
    if ($err != null) {
        return array(null, $err);
    }

    $fstat = fstat($fhandler);
    $fsize = $fstat['size'];
    $blockCnt = BlockCount($fsize);
    $sha1Buf = array();

    if ($blockCnt <= 1) {
        $sha1Buf[] = 0x16;
        list($sha1Code, $err) = CalSha1($fhandler);
        if ($err != null) {
            return array(null, $err);
        }
        fclose($fhandler);
        $sha1Buf = array_merge($sha1Buf, $sha1Code);
    } else {
        $sha1Buf[] = 0x96;
        $sha1BlockBuf = array();
        for ($i=0; $i < $blockCnt; $i++) {
            list($sha1Code, $err) = CalSha1($fhandler);
            if ($err != null) {
                return array(null, $err);
            }
            $sha1BlockBuf = array_merge($sha1BlockBuf, $sha1Code);
        }
        $tmpData = PackArray('C*', $sha1BlockBuf);
        $tmpFhandler = tmpfile();
        fwrite($tmpFhandler, $tmpData);
        fseek($tmpFhandler, 0);
        list($sha1Final, $_err) = CalSha1($tmpFhandler);
        $sha1Buf = array_merge($sha1Buf, $sha1Final);
    }
    $etag = URLSafeBase64Encode(PackArray('C*', $sha1Buf));
    return array($etag, null);
}


function isCmd () {
    $selfName = basename($_SERVER['PHP_SELF']);
    $cwd = getcwd();
    $fullPath = $cwd . '/' . $selfName;
    $isCmd = __file__ == $fullPath ? true : false;
    return $isCmd;
}


function run () {
    if ($_SERVER["argc"] < 2) {
        echo 'Usage: php qetag.php <filename>' . "\n";
        exit();
    }
    $filename = $_SERVER["argv"][1];
    list($ret, $err) = GetEtag($filename);
    if ($err != null) {
        echo $err['message'] . "\n";
        exit();
    }
    echo $ret . "\n";
    exit();
}

if (isCmd()) {
    run();
}

?>

